Background images: add defaults for background size

ramonjd opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

See: https://github.com/WordPress/gutenberg/issues/61928#issuecomment-2134387541

This PR:

- sets a default background position of `"50% 0"` for uploaded images when background size is toggled to "Tile"
- resets background position in between toggling
- defines a placeholder of "Auto" for the background image width image value.


https://github.com/WordPress/gutenberg/assets/6458278/46fc3deb-6988-45ed-9dda-bd4e92cd54b2


## Why?
To increase the chance that the image's focus point is visible.


## Testing Instructions

1. Insert a Group block and add a new background image.
2. Toggle the background size to "Tile"
3. Check that the position is set to "50% 0"
4. Toggle between size options, e.g., Cover, Contains, Tile. 
5. Check that the position resets between toggling (should it?)
6. With the size set to "Tile", upload an image or select a new one from the Media library. Check that the position is set to "50% 0"
7. Change the default position X Y values. Upload an image or select a new one from the Media library. Your position should be preserved.
8. For the Tile option, observe that the placeholder for image width input field underneath is "Auto"

Finally, to ensure that no default background position is set for theme.json-defined background images, set a background image in theme.json and toggle between the size settings in the site editor.


https://github.com/WordPress/gutenberg/assets/6458278/e7b6ec82-e7bd-442a-b498-b796a40cba05

ramonjd committed:
This commit experiments with setting a default background size of 50% for uploaded images, also saving the last known value + unit value so it can be restored between toggling

ramonjd committed:
Checking for default value when uploading/selecting an image

ramonjd committed:
Rolling back setting the default to 50% - the UX is not clear. Should we default to 50% every time the user switches to auto? What about after they deliberately clear the input?

Roll back defaultValues prop - unnecessary

ramonjd committed:
Check for an id, which indicates that the image has been uploaded to the database, and only then use `50%` as the default for Tile

jasmussen commented:
Thanks for keeping on this, and sweating the details. This is rapidly improving.

This PR solves the issue of background images by default being low resolution. But [reading up on the properties](https://www.w3schools.com/cssref/css3_pr_background-size.php), and [the back compat conversations](https://github.com/WordPress/gutenberg/pull/62000#issuecomment-2134172276), I wonder: is it possible to unset this default value if you wanted _no_ value, i.e. `auto`?

ramonjd commented:
> is it possible to unset this default value if you wanted no value, i.e. auto?

Good question! I did a quick video and added it to the description - the video suggests that yes, it's possible, but only via clearing the input field. 

I haven't tested it very thoroughly so I could be off track.

andrewserong commented:
> So: maybe we keep the tile, but add 50%/0% background position values. What do you think?

If we apply that only when a user adds an image, that sounds good to try to me! From what you've described, of all the available options, this gets as a little closer to pleasing defaults without boxing us in too much, IMO

ramonjd commented:
Thanks for all the testing @jasmussen 🙇🏻 

 I'll play around in this PR later in the week.

ramonjd committed:
When switching to "Tile" in the background size toggle, roll back default background size of 50%, and introduce a default background position of "50% 0"

ramonjd committed:
Check for existing backgroundPosition

ramonjd committed:
Clear position values between toggling

ramonjd committed:
Don't need value check as it's always reset

ramonjd commented on packages/block-editor/src/components/global-styles/background-panel.js:
```diff
@@ -426,13 +426,16 @@ function BackgroundSizeToolsPanelItem( {
 
 		if ( next === 'contain' ) {
 			nextRepeat = 'no-repeat';
+			nextPosition = undefined;
```
Setting to `undefined` between toggles so that `"50% 0"` doesn't persist.

This means however, that any user-defined position values are lost as well.

I'm wondering if we should save the value in local state and only reset when a default value is detected, e.g, `"50% 0"`

andrewserong replied:
IMO it's fine to reset the position when switching between sizes as you're kind of resetting how the image will appear anyway. Could be a follow-up PR if we think it's needed after all?

andrewserong reviewed (commented):
This is testing nicely for me when switching between background sizes on an already uploaded image. However, if I go to add an image from scratch, the default isn't added when the image is attached. I'm wondering if we also need to add a line somewhere around here in `onSelectMedia`:

https://github.com/WordPress/gutenberg/blob/f263cf8ec832289fdd676653010734780cbd5b65/packages/block-editor/src/components/global-styles/background-panel.js#L248-L255

I.e. check if `style?.background?.backgroundSize` is unset, and if so, set style.background.backgroundPosition to `50% 0`?

What do you think? I'm not sure if we've added other values when media is selected before, so wasn't sure if there's any potential edge cases to consider 🤔

jasmussen commented:
Tested this branch per the review, but I'm not seeing the defaults added when I add a background image:

![Screenshot 2024-05-30 at 08 40 41](https://github.com/WordPress/gutenberg/assets/1204802/672169d2-0c4b-43af-86dc-e04dd3e44a4f)

andrewserong commented:
Snap! Joen and I noticed the same thing at the same time 😄

jasmussen reviewed (approved):
Have to step away from this for a bit, so can't follow. But the decision about defaulting to center top position for tile, is one I can support, and this PR moves in that direction. So behavior-wise, a 👍 👍, I'll defer to Andrew and others for a code check.

andrewserong commented:
> I'll defer to Andrew and others for a code check.

Code-wise it's looking good to me. We can either land as-is, and add the behaviour when the image is first uploaded in a separate PR, or hold off on merging until that's added in. Whichever you prefer, @ramonjd!

ramonjd commented:
Thanks for the speedy reviews @jasmussen and @andrewserong 

I'll aim to add the [missing functionality](https://github.com/WordPress/gutenberg/pull/62046#pullrequestreview-2087364640) in this PR. Should be a one/two liner I think. 

Cheers!

ramonjd committed:
Ensure that the default position 50% 0 is set when changing/uploading images via the media library.

ramonjd commented:
> Tested this branch per the review, but I'm not seeing the defaults added when I add a background image:

Okay! I've added the change and updated the test instructions. 

The default of `50% 0` is set when selecting/uploading an image via the media library. 

But if the user has already set a position and not toggled from "Tile", the user value is preserved.

Thanks again!

https://github.com/WordPress/gutenberg/assets/6458278/e84040bf-ab86-4a4e-9958-dbe3e1c154f1

ramonjd commented:
Thanks for catching that final feature bug. 

I'll merge and we can carry onwards and upwards!

andrewserong commented:
Nice one, thanks for fixing that up! 🎉
